Sources/SourceKitLSP/Swift/CodeActions/ConvertStringConcatenationToStringInterpolation.swift

```diff
@@ -0,0 +1,212 @@
+  /// ``` swift
+  /// "Hello " + aString + "\(1)World"
+  /// ```
+  /// The following are invalid string concatenations.
+  /// ``` swift
+  /// aString + bString // no string literals
+  ///
+  /// "Hello " * aString - "World" // non `+` operators
+  ///
+  /// """
+  /// Hello
+  /// """
+  /// + """
+  /// World
+  /// """ // multi-line string literals
+  ///```
+  private static func preflight(
+    exprList: ExprListSyntax
+  ) -> (componentsOnly: [ExprListSyntax.Element], longestPounds: TokenSyntax?)? {
+    var iter = exprList.makeIterator()
+    guard let first = iter.next() else {
+      return nil
+    }
+
+    var hasStringComponents = false
+    var longestPounds: TokenSyntax?
+    var componentsOnly = [ExprListSyntax.Element]()
+    componentsOnly.reserveCapacity(exprList.count / 2 + 1)
+
+    if let stringLiteral = first.as(StringLiteralExprSyntax.self) {
+      guard stringLiteral.isSingleLine else {
+        return nil
+      }
+      hasStringComponents = true
+      longestPounds = stringLiteral.openingPounds
+    }
+    componentsOnly.append(first)
+
+    while let concat = iter.next(), let stringComponent = iter.next() {
+      guard let concat = concat.as(BinaryOperatorExprSyntax.self),
+        concat.operator.tokenKind == .binaryOperator("+") && !stringComponent.is(MissingExprSyntax.self)
+      else {
+        return nil
+      }
+
+      if let stringLiteral = stringComponent.as(StringLiteralExprSyntax.self) {
+        guard stringLiteral.isSingleLine else {
+          return nil
+        }
+        hasStringComponents = true
+        if let pounds = stringLiteral.openingPounds,
+          pounds.trimmedLength > (longestPounds?.trimmedLength ?? SourceLength(utf8Length: 0))
+        {
+          longestPounds = pounds
+        }
+      }
+
+      componentsOnly[componentsOnly.count - 1].trailingTrivia += concat.leadingTrivia
+      componentsOnly.append(
+        stringComponent.with(\.leadingTrivia, stringComponent.leadingTrivia + concat.trailingTrivia)
+      )
+    }
+
+    guard hasStringComponents && componentsOnly.count > 1 else {
+      return nil
+    }
+
+    return (componentsOnly, longestPounds)
+  }
+}
+
+extension ConvertStringConcatenationToStringInterpolation: SyntaxRefactoringCodeActionProvider {
+  static let title: String = "Convert String Concatenation to String Interpolation"
+
+  static func nodeToRefactor(in scope: SyntaxCodeActionScope) -> SequenceExprSyntax? {
+    guard let expr = scope.innermostNodeContainingRange,
+      let seqExpr = expr.findParentOfSelf(
+        ofType: SequenceExprSyntax.self,
+        stoppingIf: {
+          $0.kind == .codeBlockItem || $0.kind == .memberBlockItem
+        }
+      )
+    else {
+      return nil
+    }
+
+    return seqExpr
+  }
+}
+
+fileprivate extension StringLiteralExprSyntax {
+  var isSingleLine: Bool {
+    openingQuote.tokenKind == .stringQuote
+  }
+}
+
+fileprivate extension SyntaxProtocol {
+  var commentsInlined: Self {
```

Comment:
I think the primary action here is that the trivia gets flattened to a single line.

```suggestion
  /// Modifies the trivia to not contain any newlines. This removes whitespace trivia, removes newlines from block comments and converts line comments to block comments.
  var withSingleLineTrivia: Self {
```